File: frontend/lib/print-forms.ts
/**
 * Print-form registry helpers (Stage 18.3.5 frontend).
 */

export type PrintFormBindingType = "model" | "directory" | "document_type";
export type PrintFormStatus = "draft" | "active" | "archived";
export type PrintFormOutputFormat = "html" | "pdf" | "xlsx";
export type PrintFormVersionStatus = "draft" | "published" | "archived";
export type PrintFormVersionStorageKind = "inline_text" | "file_ref";

export type PrintFormVersion = {
  id: number;
  print_form_id: number;
  version_no: number;
  template_label: string;
  storage_kind: string;
  template_source: string;
  status: string;
  is_current: boolean;
  created_at: string;
  updated_at: string;
};

export type PrintForm = {
  id: number;
  code: string;
  title: string;
  description: string | null;
  binding_type: string;
  binding_key: string;
  status: string;
  output_format: string;
  versioning_mode: string;
  created_at: string;
  updated_at: string;
  versions: PrintFormVersion[];
};

export type PrintFormDraft = {
  code: string;
  title: string;
  description: string;
  binding_type: PrintFormBindingType;
  binding_key: string;
  output_format: PrintFormOutputFormat;
};

export type PrintFormVersionDraft = {
  template_label: string;
  storage_kind: PrintFormVersionStorageKind;
  template_source: string;
};

export type PrintFormPreview = {
  print_form_id: number;
  print_form_code: string;
  version_id: number;
  version_no: number;
  output_format: string;
  content_type: string;
  file_name: string;
  content: string;
  is_preview: boolean;
};

export const printFormBindingTypeOptions: Array<{
  value: PrintFormBindingType;
  label: string;
}> = [
  { value: "model", label: "Модель" },
  { value: "directory", label: "Справочник" },
  { value: "document_type", label: "Тип документа" },
];

export const printFormOutputFormatOptions: Array<{
  value: PrintFormOutputFormat;
  label: string;
}> = [
  { value: "html", label: "HTML" },
  { value: "pdf", label: "PDF" },
  { value: "xlsx", label: "XLSX" },
];

export const printFormVersionStatusLabels: Record<string, string> = {
  draft: "Черновик",
  published: "Опубликована",
  archived: "Архив",
};

export const printFormStatusLabels: Record<string, string> = {
  draft: "Черновик",
  active: "Активна",
  archived: "Архив",
};

export function emptyPrintFormDraft(): PrintFormDraft {
  return {
    code: "",
    title: "",
    description: "",
    binding_type: "model",
    binding_key: "",
    output_format: "html",
  };
}

export function printFormToDraft(printForm: PrintForm): PrintFormDraft {
  return {
    code: printForm.code,
    title: printForm.title,
    description: printForm.description ?? "",
    binding_type: normalizeBindingType(printForm.binding_type),
    binding_key: printForm.binding_key,
    output_format: normalizeOutputFormat(printForm.output_format),
  };
}

export function emptyPrintFormVersionDraft(): PrintFormVersionDraft {
  return {
    template_label: "",
    storage_kind: "inline_text",
    template_source: "",
  };
}

export function validatePrintFormDraft(draft: PrintFormDraft): string | null {
  if (!draft.code.trim()) return "Укажите код печатной формы";
  if (!/^[a-z0-9_]+$/.test(draft.code.trim())) {
    return "Код должен содержать только латиницу, цифры и подчёркивания";
  }
  if (!draft.title.trim()) return "Укажите название печатной формы";
  if (!draft.binding_key.trim()) return "Укажите ключ привязки";
  if (!/^[a-z0-9_]+$/.test(draft.binding_key.trim())) {
    return "Ключ привязки должен содержать только латиницу, цифры и подчёркивания";
  }
  if (draft.title.trim().length > 160) {
    return "Название не длиннее 160 символов";
  }
  if (draft.description.trim().length > 2000) {
    return "Описание не длиннее 2000 символов";
  }
  return null;
}

export function validatePrintFormVersionDraft(
  draft: PrintFormVersionDraft,
): string | null {
  if (!draft.template_label.trim()) return "Укажите метку версии";
  if (!draft.template_source.trim()) return "Укажите текст шаблона";
  if (draft.template_label.trim().length > 160) {
    return "Метка версии не длиннее 160 символов";
  }
  return null;
}

export function filterPrintForms(
  rows: PrintForm[],
  query: string,
  activeOnly: boolean,
): PrintForm[] {
  const needle = query.trim().toLowerCase();
  return rows.filter((row) => {
    if (activeOnly && row.status !== "active") return false;
    if (!needle) return true;
    return (
      row.code.toLowerCase().includes(needle) ||
      row.title.toLowerCase().includes(needle) ||
      row.binding_key.toLowerCase().includes(needle)
    );
  });
}

export function getCurrentVersion(
  printForm: PrintForm,
): PrintFormVersion | null {
  return printForm.versions.find((version) => version.is_current) ?? null;
}

function normalizeBindingType(value: string): PrintFormBindingType {
  if (value === "directory" || value === "document_type") return value;
  return "model";
}

function normalizeOutputFormat(value: string): PrintFormOutputFormat {
  if (value === "pdf" || value === "xlsx") return value;
  return "html";
}

File: backend/app/schemas/print_forms.py
"""Print-form registry schemas (18.3.2)."""

from __future__ import annotations

from typing import Any
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class PrintFormVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    print_form_id: int
    version_no: int
    template_label: str
    storage_kind: str
    template_source: str
    status: str
    is_current: bool
    created_at: datetime
    updated_at: datetime


class PrintFormRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    title: str
    description: str | None = None
    binding_type: str
    binding_key: str
    status: str
    output_format: str
    versioning_mode: str
    created_at: datetime
    updated_at: datetime
    versions: list[PrintFormVersionRead] = []


class PrintFormCreate(BaseModel):
    code: str = Field(min_length=1, max_length=80)
    title: str = Field(min_length=1, max_length=160)
    description: str | None = None
    binding_type: str = Field(min_length=1, max_length=20)
    binding_key: str = Field(min_length=1, max_length=120)
    status: str = Field(default="draft", min_length=1, max_length=20)
    output_format: str = Field(default="html", min_length=1, max_length=20)
    versioning_mode: str = Field(
        default="single_active",
        min_length=1,
        max_length=20,
    )


class PrintFormUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = None
    binding_type: str | None = Field(default=None, min_length=1, max_length=20)
    binding_key: str | None = Field(default=None, min_length=1, max_length=120)
    status: str | None = Field(default=None, min_length=1, max_length=20)
    output_format: str | None = Field(default=None, min_length=1, max_length=20)
    versioning_mode: str | None = Field(default=None, min_length=1, max_length=20)


class PrintFormVersionCreate(BaseModel):
    template_label: str = Field(min_length=1, max_length=160)
    storage_kind: str = Field(default="inline_text", min_length=1, max_length=20)
    template_source: str = Field(min_length=1)
    status: str = Field(default="draft", min_length=1, max_length=20)
    is_current: bool = False


class PrintFormVersionUpdate(BaseModel):
    template_label: str | None = Field(default=None, min_length=1, max_length=160)
    storage_kind: str | None = Field(default=None, min_length=1, max_length=20)
    template_source: str | None = Field(default=None, min_length=1)
    status: str | None = Field(default=None, min_length=1, max_length=20)
    is_current: bool | None = None


class PrintFormVersionPublishRequest(BaseModel):
    is_current: bool = True


class PrintFormPreviewRequest(BaseModel):
    version_id: int | None = None
    payload: dict[str, Any] = Field(default_factory=dict)


class PrintFormGenerateRequest(BaseModel):
    binding_type: str = Field(min_length=1, max_length=20)
    binding_key: str = Field(min_length=1, max_length=120)
    output_format: str = Field(default="html", min_length=1, max_length=20)
    payload: dict[str, Any] = Field(default_factory=dict)


class PrintFormRenderRead(BaseModel):
    print_form_id: int
    print_form_code: str
    version_id: int
    version_no: int
    output_format: str
    content_type: str
    file_name: str
    content: str
    content_encoding: str = "text"
    is_preview: bool
